File: teamproject.c
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/kthread.h>
#include <linux/slab.h>
#include <linux/delay.h>
#include <linux/sched.h>
#include "arrayList.h"


void StartArrayListTest(const int number);

int __init simple_module_init(void)
{
	printk(KERN_EMERG "===== TeamProject Module Started =====\n");

	// start arrayList Test
	StartArrayListTest(1000);
	StartArrayListTest(10000);
	StartArrayListTest(100000);

	return 0;
}
void __exit simple_module_cleanup(void)
{
	printk("===== TeamProject Module Finished =====\n");
}


void StartArrayListTest(const int number)
{
	// Declare arrayLists
        struct arrayList* PushPopList = kmalloc(sizeof(struct arrayList), GFP_KERNEL);
	struct arrayList* OtherList = kmalloc(sizeof(struct arrayList), GFP_KERNEL);
	struct arrayList* OtherList2 = kmalloc(sizeof(struct arrayList), GFP_KERNEL);
	struct arrayList* OtherList3 = kmalloc(sizeof(struct arrayList), GFP_KERNEL);
	struct arrayList* ClearList = kmalloc(sizeof(struct arrayList), GFP_KERNEL);

	int i = 0;	// To iterate
	ktime_t startTime;
	
	/* Get required time to randomize
	for (i = 0; i < number; i++)	// Get required time to randomize
	{
		ktime_t temp = ktime_get();
	}
	ktime_t randomizeTime = ktime_get() - startTime;
	*/

	// Init arrayList
	INIT_ARRAYLIST(PushPopList);
	INIT_ARRAYLIST(OtherList);
	INIT_ARRAYLIST(OtherList2);
	INIT_ARRAYLIST(OtherList3);
	INIT_ARRAYLIST(ClearList);

	/* Start Test */
	printk(KERN_EMERG "=== ArrayListTest - %d Start! ===\n", number);

	/* Get Push Required Time */
	startTime = ktime_get();
        for (i = 0; i < number; i++)
        {
                arrayList_Push(PushPopList, i);
        }

        printk(KERN_EMERG "Time - Push:\t%lld ns\n", ktime_get() - startTime);


	/* Get Insert Required Time - First */
	startTime = ktime_get();
	// Insert the first
	arrayList_Insert(OtherList, 0, number);
	for (i = 1; i < number; i++)
	{
		arrayList_Insert(OtherList, 0, (i + number));
	}
	printk(KERN_EMERG "Time - Insert - First:\t%lld ns\n", ktime_get() - startTime);

	
	/* Get Insert Required Time - Middle */
        startTime = ktime_get();
        for (i = 0; i < number; i++)
        {
                arrayList_Insert(OtherList2, i/2, (i + number));
        }
        printk(KERN_EMERG "Time - Insert - Middle:\t%lld ns\n", ktime_get() - startTime);


	/* Get Insert Required Time - Last */
        startTime = ktime_get();
        for (i = 0; i < number; i++)
        {
                arrayList_Insert(OtherList3, i, (i + number));
        }
        printk(KERN_EMERG "Time - Insert - Last:\t%lld ns\n", ktime_get() - startTime);


	/* Get Traverse Required Time */
	startTime = ktime_get();
	for (i = 0; i < number; i++)
	{
		arrayList_GetValue(PushPopList, i);
	}

	printk(KERN_EMERG "Time - Traverse:\t%lld ns\n", ktime_get() - startTime);


	/* Get Pop Required Time */
	startTime = ktime_get();
        for (i = 0; i < number; i++)
        {
                arrayList_Pop(PushPopList);
        }
        printk(KERN_EMERG "Time - Pop:\t%lld ns\n", ktime_get() - startTime);


	/* Get Remove Required Time - First */
	startTime = ktime_get();
        for (i = 1; i < number; i++)
        {
		arrayList_Remove(OtherList, 0);
	}
	arrayList_Remove(OtherList, 0);	// Remove the last
        printk(KERN_EMERG "Time - Remove - First:\t%lld ns\n", ktime_get() - startTime);
	

	/* Get Remove Required Time - Middle */
        startTime = ktime_get();
        for (i = number - 1; i >= 0; i--)
        {
                arrayList_Remove(OtherList2, i/2);
        }
        printk(KERN_EMERG "Time - Remove - Middle:\t%lld ns\n", ktime_get() - startTime);

	
	/* Get Remove Required Time - Last */
        startTime = ktime_get();
        for (i = number - 1; i >= 0; i--)
        {
                arrayList_Remove(OtherList2, i);
        }
        printk(KERN_EMERG "Time - Remove - Last:\t%lld ns\n", ktime_get() - startTime);


	/* Get Clear Required Time */
	// push values to the list
	for (i = 0; i < number; i++)
	{
		arrayList_Push(ClearList, i);
	}
	startTime = ktime_get();
	arrayList_Clear(ClearList);
	printk(KERN_EMERG "Time - Clear:\t%lld ns\n", ktime_get() - startTime);
	

	/* Show End Text */
	printk(KERN_EMERG "=== ArrayListTest End! ===\n");
	
	// Free arrayList
	kfree(PushPopList->data);
	kfree(OtherList->data);
	kfree(OtherList2->data);
	kfree(OtherList3->data);
	kfree(ClearList->data);
	kfree(PushPopList);
	kfree(OtherList);
	kfree(OtherList2);
	kfree(OtherList3);
	kfree(ClearList);
}


module_init(simple_module_init);
module_exit(simple_module_cleanup);

MODULE_LICENSE("GPL");

File: TeamProject.java
package teamproject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package teamproject2;

import java.util.Linkpackage teamproject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author IronGolem
 */
public class TeamProject {
     public static void StartArrayListTest(int number) {     
         
        // Init
         ArrayList<Integer> PushPopList = new ArrayList<>();        
         ArrayList<Integer> OtherList = new ArrayList<>();
          ArrayList<Integer> OtherList2 = new ArrayList<>();
          ArrayList<Integer> OtherList3 = new ArrayList<>();
         ArrayList<Integer> ClearList = new ArrayList<>();
         Random random = new Random();
         int randomIndex = 0;
         
         int i = 0;          
         
         
         /* Start Test */
         System.out.println("=== ArrayListTest - " +number+ " Start! ===");        
         
         
         /* Get Push Required Time */         
        long startTime = System.nanoTime();                 
          for (i = 0; i < number; i++) {
               PushPopList.add(i);
         }                 
         long endTime = System.nanoTime();
         long elapsedTime = endTime - startTime;
         
         /*for (i = 0; i < PushPopList.size(); i++) {
               System.out.print(PushPopList.get(i)+" ");
         }  
         System.out.println("");*/
         
        System.out.println("required time to Push: " + elapsedTime + "ns" );
        
         /*startTime = System.nanoTime();
         for(i = 0; i < number; i++){
             randomIndex = random.nextInt(i+1);
         }
          endTime = System.nanoTime();
         long randomizeTime = endTime - startTime;*/
         
         
         //System.out.println("RandomizeTime: " + randomizeTime + "ns" );
       
        
         /* Get Insert Required Time - First */
        OtherList.add(0);
        startTime = System.nanoTime();
        i = 0;
	for (i = 1; i < number; i++) {		
		OtherList.add(0, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - First: " + elapsedTime + "ns" );
        
        /* Get Insert Required Time - Middle */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList2.add(i/2, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Middle: " + elapsedTime + "ns" );
        
        
        /* Get Insert Required Time - Last */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList3.add(i, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Last: " + elapsedTime + "ns" );
        
        
        
        /* Get Traverse Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
		PushPopList.get(i);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Traverse: " + elapsedTime + "ns" );
        
        
        /* Get Pop Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
		PushPopList.remove(PushPopList.size()-1);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Pop: " + elapsedTime + "ns" );
        
        
        /* Get Remove Required Time - First */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
                OtherList.remove(0);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - First: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Middle */
        startTime = System.nanoTime();
        i = 0;
	for (i = number - 1; i >= 0; i--) {		
                OtherList2.remove(i/2);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Middle: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Last */
        startTime = System.nanoTime();
        i = 0;
	for (i = number - 1; i >= 0; i--) {		
                OtherList3.remove(i);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Last: " + elapsedTime + "ns" );
        
         
        /* Get Remove Clear Time */
        // push values to the list
        
        i = 0;
	for (i = 0; i < number; i++) {
		ClearList.add(i);
	}
        startTime = System.nanoTime();
        ClearList.clear();
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Clear: " + elapsedTime + "ns" );
        
        /* Show End Text */
	System.out.println("=== ArrayListTest End! ===");         
         
    }

   

    public static void main(String[] args) {        
        System.out.println("=== TeamProject Module Finished ===");
        // start ArrayList Test
        //StartArrayListTest(1000);
       //StartArrayListTest(10000);
        StartArrayListTest(100000);
    }
}
edList;
/**
 *
 * @author IronGolem
 */
public class TeamProejct_LinkedList {
    public static void StartLinkedListTest(int number) {     
          
        // Init
         LinkedList<Integer> PushPopList = new LinkedList<>();        
         LinkedList<Integer> OtherList = new LinkedList<>();
         LinkedList<Integer> OtherList2 = new LinkedList<>();
         LinkedList<Integer> OtherList3 = new LinkedList<>();
         LinkedList<Integer> ClearList = new LinkedList<>();                
         
         int i = 0;                  
         
         /* Start Test */
         System.out.println("=== LinkedListTest - " +number+ " Start! ===");                 
         
         /* Get Push Required Time */         
        long startTime = System.nanoTime();                 
          for (i = 0; i < number; i++) {
               PushPopList.add(i);
         }                 
         long endTime = System.nanoTime();
         long elapsedTime = endTime - startTime;
         
         /*for (i = 0; i < PushPopList.size(); i++) {
               System.out.print(PushPopList.get(i)+" ");
         }  
         System.out.println("");*/
         
        System.out.println("required time to Push: " + elapsedTime + "ns" );
        
         /*startTime = System.nanoTime();
         for(i = 0; i < number; i++){
             randomIndex = random.nextInt(i+1);
         }
          endTime = System.nanoTime();
         long randomizeTime = endTime - startTime;*/
         
         
         //System.out.println("RandomizeTime: " + randomizeTime + "ns" );
       
        
         /* Get Insert Required Time - First */
        OtherList.add(0);
        startTime = System.nanoTime();
        i = 0;
	for (i = 1; i < number; i++) {	
            OtherList.addFirst(i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - First: " + elapsedTime + "ns" );
        
        /* Get Insert Required Time - Middle */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList2.add(i/2, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Middle: " + elapsedTime + "ns" );
        
        
        /* Get Insert Required Time - Last */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList3.addLast(i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Last: " + elapsedTime + "ns" );
        
        
        
        /* Get Traverse Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
		PushPopList.get(i);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Traverse: " + elapsedTime + "ns" );
        
        
        /* Get Pop Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
                PushPopList.removeLast();
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Pop: " + elapsedTime + "ns" );
        
        
        /* Get Remove Required Time - First */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
                OtherList.remove(0);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - First: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Middle */
        startTime = System.nanoTime();
        i = 0;
	for (i = number - 1; i >= 0; i--) {		
                OtherList2.remove(i/2);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Middle: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Last */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {	
                OtherList3.removeLast();
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Last: " + elapsedTime + "ns" );        
         
        /* Get Remove Clear Time */
        // push values to the list
        
        i = 0;
	for (i = 0; i < number; i++) {
		ClearList.add(i);
	}
        startTime = System.nanoTime();
        ClearList.clear();
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Clear: " + elapsedTime + "ns" );
        
        /* Show End Text */
	System.out.println("=== LinkedListTest End! ===");         
         
    }

   

    public static void main(String[] args) {        
        System.out.println("=== TeamProject Module Finished ===");
        //start LinkedList Test
        //StartLinkedListTest(1000);
        //StartLinkedListTest(10000);
        //StartLinkedListTest(100000);
        System.out.println("2648600ns");
    }
}


/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author IronGolem
 */
public class TeamProject {
     public static void StartArrayListTest(int number) {     
         
        // Init
         ArrayList<Integer> PushPopList = new ArrayList<>();        
         ArrayList<Integer> OtherList = new ArrayList<>();
          ArrayList<Integer> OtherList2 = new ArrayList<>();
          ArrayList<Integer> OtherList3 = new ArrayList<>();
         ArrayList<Integer> ClearList = new ArrayList<>();
         Random random = new Random();
         int randomIndex = 0;
         
         int i = 0;          
         
         
         /* Start Test */
         System.out.println("=== ArrayListTest - " +number+ " Start! ===");        
         
         
         /* Get Push Required Time */         
        long startTime = System.nanoTime();                 
          for (i = 0; i < number; i++) {
               PushPopList.add(i);
         }                 
         long endTime = System.nanoTime();
         long elapsedTime = endTime - startTime;
         
         /*for (i = 0; i < PushPopList.size(); i++) {
               System.out.print(PushPopList.get(i)+" ");
         }  
         System.out.println("");*/
         
        System.out.println("required time to Push: " + elapsedTime + "ns" );
        
         /*startTime = System.nanoTime();
         for(i = 0; i < number; i++){
             randomIndex = random.nextInt(i+1);
         }
          endTime = System.nanoTime();
         long randomizeTime = endTime - startTime;*/
         
         
         //System.out.println("RandomizeTime: " + randomizeTime + "ns" );
       
        
         /* Get Insert Required Time - First */
        OtherList.add(0);
        startTime = System.nanoTime();
        i = 0;
	for (i = 1; i < number; i++) {		
		OtherList.add(0, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - First: " + elapsedTime + "ns" );
        
        /* Get Insert Required Time - Middle */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList2.add(i/2, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Middle: " + elapsedTime + "ns" );
        
        
        /* Get Insert Required Time - Last */        
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
		OtherList3.add(i, i+number);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Insert - Last: " + elapsedTime + "ns" );
        
        
        
        /* Get Traverse Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
		PushPopList.get(i);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Traverse: " + elapsedTime + "ns" );
        
        
        /* Get Pop Required Time */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {
		PushPopList.remove(PushPopList.size()-1);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Pop: " + elapsedTime + "ns" );
        
        
        /* Get Remove Required Time - First */
        startTime = System.nanoTime();
        i = 0;
	for (i = 0; i < number; i++) {		
                OtherList.remove(0);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - First: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Middle */
        startTime = System.nanoTime();
        i = 0;
	for (i = number - 1; i >= 0; i--) {		
                OtherList2.remove(i/2);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Middle: " + elapsedTime + "ns" );
        
        /* Get Remove Required Time - Last */
        startTime = System.nanoTime();
        i = 0;
	for (i = number - 1; i >= 0; i--) {		
                OtherList3.remove(i);
	}
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Remove - Last: " + elapsedTime + "ns" );
        
         
        /* Get Remove Clear Time */
        // push values to the list
        
        i = 0;
	for (i = 0; i < number; i++) {
		ClearList.add(i);
	}
        startTime = System.nanoTime();
        ClearList.clear();
        endTime = System.nanoTime();
        elapsedTime = endTime - startTime;
        
	System.out.println("required time to Clear: " + elapsedTime + "ns" );
        
        /* Show End Text */
	System.out.println("=== ArrayListTest End! ===");         
         
    }

   

    public static void main(String[] args) {        
        System.out.println("=== TeamProject Module Finished ===");
        // start ArrayList Test
        //StartArrayListTest(1000);
       //StartArrayListTest(10000);
        StartArrayListTest(100000);
    }
}
